# SQL Part 3 Studio

### Let's practice your SQL quering skills!  For each question, work along in the notebook adding your query and answering the questions. 

### **DataSet**

### <span style="font-size: 14px; font-family: -apple-system, BlinkMacSystemFont, sans-serif;">We will be using the Goodbooks-10k dataset from the exercises in the prep work.&nbsp; Feel free reference your previous notebooks.</span>

- The dataset can be found here: [goodbooks-10k](https://www.kaggle.com/zygmunt/goodbooks-10k)
- You can access the **BooksDB** in the Launchcode server.

### **Business Issue**

You are an employee at a small independet bookstore and you have been tasked to increase sales.  You decide to use the BooksDB to find books and themes to highlight in fun promotions throughout each month / season.  We will be starting with March and then you will have a chance to explore another month of your choosing.  We want to come up with a list of events/ promotions to run each month.  If you are looking for ideas to include in your promotions:

- [https://www.calendarr.com/united-states/calendar-2022/](https://www.calendarr.com/united-states/calendar-2022/)
- [https://www.holidayinsights.com/moreholidays/](https://www.holidayinsights.com/moreholidays/)
    - Click on a month and it will take you to a more detailed page

## Part 1:  March - Women's History Month, Dr Seuss Day(3/2), National Pie Day3/14, St Patrick's Day (3/17), Season - Spring

### **Event 1:** Women's History Month

Goal: Highlight Popular Womans Authors based on ratings from the Good Reads DB.

### **Write a query that returns tag\_id, the number of times each tag\_id is used and the tag\_name.**  

### Use the GROUP BY and HAVING clause to narrow your focus.

Try multiple keywords, ie woman vs female (female ended up giving better results)

In [49]:
-- Solution
select * from BooksDB.dbo.tags where tag_name like '%woman%' or tag_name like '%female%';

select tag_id from BooksDB.dbo.tags where tag_name like 'author-female' or tag_name like 'female-author%' or tag_name like 'female-writer&' or tag_name like 'woman-author' or tag_name like 'woman-trans-agender-authors';

select t.tag_name, sum(t.tag_id) as TagID, bt.count
from BooksDB.dbo.book_tags as bt
left join BooksDB.dbo.tags as t
on t.tag_id = bt.tag_id
group by t.tag_name, t.tag_id, bt.count
having t.tag_name like 'author-female' or t.tag_name like 'female-author%' or t.tag_name like 'female-writer&' or t.tag_name like 'woman-author' or t.tag_name like 'woman-trans-agender-authors'
order by t.tag_name asc;

(59 rows affected)

(5 rows affected)

(173 rows affected)

Total execution time: 00:00:00.737

tag_id tag_name 937 21-books-every-woman-should-read 1386 a-woman-after-god-s-own-heart 1387 a-woman-after-gods-own-heart 1741 afghan-female-entrepreneur-under-ta 3514 author-female 4006 batwoman 4309 best-adult-female-friendship-books 4455 biblical-womanhood 4729 black-ops-woman-team-cia-sean-case 5132 books-every-woman-should-read 7194 christian-woman 10986 every-woman-should-read 11642 female 11643 female-author 11644 female-authors 11645 female-biographies 11646 female-detective 11647 female-detectives 11648 female-doctor 11649 female-fiction 11650 female-friendship 11651 female-genre-authors 11652 female-lead 11653 female-main-character 11654 female-persuasion 11655 female-pi 11656 female-pov 11657 female-povs 11658 female-protagonist 11659 female-serial-killer 11660 female-spy 11661 female-writer 11662 female-writers 15331 important-female-authors 17079 kick-ass-female 19038 man-woman 19757 memorable-woman-characters 22269 older-man-younger-woman 22270 older-woman 22271 older-woman-younger-man 23671 pioneer-woman 23891 policewoman-potential-victim 24455 protagonist-female 28830 strong-female 28831 strong-female-character 28832 strong-female-character-s 28833 strong-female-characters 28834 strong-female-lead 28835 strong-female-protagonist 28836 strong-female-supporting-characters 29324 tbr-ax-female 32679 woman-author 32680 woman-s-fiction 32681 woman-s-murder-club 32682 woman-trans-agender-authors 32683 womanhood 32684 womans-literature 32685 womans-murder-club 32740 wonder-woman

tag_id 3514 11643 11644 32679 32682

tag_name TagID count author-female 3514 9 author-female 3514 4 author-female 3514 10 author-female 17570 2 author-female 10542 3 author-female 3514 5 female-author 698580 5 female-author 11643 82 female-author 46572 60 female-author 46572 68 female-author 23286 75 female-author 93144 23 female-author 11643 77 female-author 34929 42 female-author 151359 28 female-author 360933 14 female-author 11643 83 female-author 104787 24 female-author 11643 91 female-author 279432 2 female-author 34929 46 female-author 430791 10 female-author 11643 38 female-author 23286 51 female-author 605436 9 female-author 58215 39 female-author 593793 3 female-author 93144 37 female-author 326004 16 female-author 11643 150 female-author 46572 54 female-author 11643 87 female-author 34929 40 female-author 23286 47 female-author 46572 33 female-author 558864 8 female-author 23286 61 female-author 128073 26 female-author 46572 1 female-author 232860 17 female-author 337647 15 female-author 267789 11 female-author 69858 34 female-author 314361 22 female-author 128073 25 female-author 58215 48 female-author 11643 36 female-author 430791 12 female-author 23286 64 female-author 34929 70 female-author 11643 63 female-author 11643 50 female-author 11643 198 female-author 58215 44 female-author 128073 31 female-author 11643 108 female-author 46572 53 female-author 11643 66 female-author 93144 30 female-author 174645 21 female-author 617079 7 female-author 11643 65 female-author 58215 35 female-author 244503 19 female-author 768438 6 female-author 34929 43 female-author 11643 55 female-author 34929 76 female-author 23286 57 female-author 11643 137 female-author 34929 59 female-author 11643 58 female-author 11643 79 female-author 34929 49 female-author 11643 67 female-author 11643 103 female-author 465720 4 female-author 11643 140 female-author 11643 93 female-author 512292 13 female-author 11643 104 female-author 81501 41 female-author 384219 18 female-author 11643 194 female-author 197931 32 female-author 116430 27 female-author 232860 20 female-author 46572 45 female-author 128073 29 female-authors 46576 23 female-authors 81508 30 female-authors 23288 66 female-authors 11644 106 female-authors 11644 53 female-authors 34932 60 female-authors 465760 5 female-authors 11644 49 female-authors 34932 62 female-authors 11644 158 female-authors 11644 214 female-authors 11644 83 female-authors 23288 75 female-authors 104796 29 female-authors 23288 76 female-authors 11644 102 female-authors 104796 24 female-authors 326032 16 female-authors 81508 26 female-authors 23288 89 female-authors 256168 2 female-authors 11644 128 female-authors 395896 10 female-authors 11644 44 female-authors 11644 47 female-authors 81508 33 female-authors 384252 11 female-authors 34932 63 female-authors 23288 93 female-authors 104796 20 female-authors 23288 34 female-authors 11644 72 female-authors 69864 41 female-authors 407540 4 female-authors 34932 25 female-authors 558912 6 female-authors 23288 48 female-authors 69864 37 female-authors 11644 54 female-authors 81508 27 female-authors 46576 35 female-authors 58220 40 female-authors 23288 51 female-authors 256168 13 female-authors 267812 18 female-authors 46576 58 female-authors 23288 32 female-authors 360964 9 female-authors 314388 3 female-authors 11644 81 female-authors 11644 170 female-authors 116440 22 female-authors 11644 90 female-authors 11644 50 female-authors 163016 17 female-authors 23288 64 female-authors 11644 84 female-authors 69864 43 female-authors 232880 15 female-authors 11644 67 female-authors 34932 56 female-authors 23288 57 female-authors 11644 91 female-authors 34932 1 female-authors 454116 8 female-authors 11644 82 female-authors 23288 52 female-authors 11644 71 female-authors 81508 36 female-authors 11644 59 female-authors 197948 19 female-authors 104796 21 female-authors 11644 146 female-authors 174660 14 female-authors 139728 31 female-authors 23288 38 female-authors 58220 28 female-authors 34932 39 female-authors 34932

### **Which keyword(s) and tag(s) would you like to explore more?**

author-female, female-author, female-authors, female-writer, <span style="color: var(--vscode-foreground);">female-writers, woman-author, and woman-trans-agender-authors</span><span style="color: var(--vscode-foreground);">.</span>

### **Write a query to return authors, titles, ratings and** **selected tag(s)\_id.**

In [56]:
-- Solution
-- select * from BooksDB.dbo.tags where tag_name like 'strong-female-lead' or tag_name like 'female-friendship';

select b.original_title as "Title", b.authors as "Authors", b.average_rating as "Average Rating", bt.tag_id as "Tag ID"
from BooksDB.dbo.book_tags as bt
inner join BooksDB.dbo.books as b
on b.best_book_id = bt.goodreads_book_id
where b.original_title is not null
group by b.original_title,b.authors,b.average_rating,bt.tag_id
having bt.tag_id like '3514' or bt.tag_id like '11643' or bt.tag_id like '11644' or bt.tag_id like '32679' or bt.tag_id like '32682'
order by b.average_rating desc;

(1628 rows affected)

Total execution time: 00:00:01.189

Title Authors Average Rating Tag ID Harry Potter Boxed Set Books 1-4 J.K. Rowling, Mary GrandPré 4.66 11643 桜蘭高校ホスト部 15 Bisco Hatori 4.54 11643 Fool's Quest Robin Hobb 4.53 11643 Fool's Quest Robin Hobb 4.53 11644 Kings Rising C.S. Pacat 4.5 11644 The New Jim Crow: Mass Incarceration in the Age of Colorblindness Michelle Alexander 4.49 11643 鋼の錬金術師 1 Hiromu Arakawa, Akira Watanabe 4.49 11644 The Gruffalo Julia Donaldson, Axel Scheffler 4.47 11643 We Should All Be Feminists Chimamanda Ngozi Adichie 4.47 11643 We Should All Be Feminists Chimamanda Ngozi Adichie 4.47 11644 Harry Potter and the Order of the Phoenix J.K. Rowling, Mary GrandPré 4.46 11643 The Complete Stories Flannery O'Connor 4.44 11643 The Complete Stories Flannery O'Connor 4.44 11644 A Civil Campaign Lois McMaster Bujold 4.43 11643 Nothing to Envy: Ordinary Lives in North Korea Barbara Demick 4.42 11644 Homegoing Yaa Gyasi 4.41 11643 Homegoing Yaa Gyasi 4.41 11644 The Sunne in Splendour Sharon Kay Penman 4.4 11643 I Love You Through And Through Bernadette Rossetti-Shustak, Caroline Jayne Church 4.39 11644 Storia della bambina perduta Elena Ferrante, Ann Goldstein 4.39 11643 Storia della bambina perduta Elena Ferrante, Ann Goldstein 4.39 11644 Fool's Fate Robin Hobb 4.38 11643 Tales From a Not-So-Graceful Ice Princess Rachel Renée Russell 4.38 11643 Kushiel's Avatar Jacqueline Carey 4.37 11643 Queen of the Darkness Anne Bishop 4.37 11643 Storia del nuovo cognome Elena Ferrante, Ann Goldstein 4.37 11643 Storia del nuovo cognome Elena Ferrante, Ann Goldstein 4.37 11644 The King of Attolia Megan Whalen Turner 4.37 11643 Persepolis Marjane Satrapi, Mattias Ripa, Blake Ferris, Anjali Singh 4.36 11643 Persepolis Marjane Satrapi, Mattias Ripa, Blake Ferris, Anjali Singh 4.36 11644 The Obelisk Gate N.K. Jemisin 4.36 11643 Bröderna Lejonhjärta Astrid Lindgren, Ilon Wikland, Jill M. Morgan 4.35 11643 Fool's Assassin Robin Hobb 4.35 11643 Sunset Erin Hunter 4.35 11643 The Scar of David Susan Abulhawa 4.35 11644 A Gathering of Shadows V.E. Schwab, Victoria Schwab 4.33 11643 Amazing Grace Mary Hoffman, Caroline Binch 4.33 11643 Mastering the Art of French Cooking Julia Child, Simone Beck, Louisette Bertholle 4.33 11644 Mirror Dance Lois McMaster Bujold 4.32 11643 The Folk of Faraway Tree Enid Blyton 4.32 11644 The Golden Fool Robin Hobb 4.32 11643 The Realms of the Gods (Immortals, #4) Tamora Pierce 4.32 11643 Wild Magic (Immortals, #1) Tamora Pierce 4.32 11643 Wild Magic (Immortals, #1) Tamora Pierce 4.32 11644 Women's Bodies, Women's Wisdom: Creating Physical and Emotional Health and Healing Christiane Northrup 4.32 11643 A Kiss for Little Bear Else Holmelund Minarik, Maurice Sendak 4.31 11643 Dawn Erin Hunter 4.31 11643 Kushiel's Chosen Jacqueline Carey 4.31 11643 Moonrise Erin Hunter 4.31 11643 The Fifth Season N.K. Jemisin 4.31 11643 The Fifth Season N.K. Jemisin 4.31 11644 The Warmth of Other Suns: The Epic Story of America's Great Migration Isabel Wilkerson 4.31 11643 The Way to Cook Julia Child 4.31 11643 The Way to Cook Julia Child 4.31 11644 Citizen: An American Lyric Claudia Rankine 4.3 11643 Citizen: An American Lyric Claudia Rankine 4.3 11644 Emperor Mage (Immortals, #3) Tamora Pierce 4.3 11643 Here be Dragons Sharon Kay Penman 4.3 11643 Howl's Moving Castle Diana Wynne Jones 4.3 11643 Howl's Moving Castle Diana Wynne Jones 4.3 11644 Kushiel's Mercy Jacqueline Carey 4.3 11643 Lady Knight Tamora Pierce 4.3 11643 Rudolph the Red-Nosed Reindeer Barbara Shook Hazen, Robert Lewis May, Richard Scarry 4.3 11644 The Places That Scare You: A Guide to Fearlessness in Difficult Times Pema Chödrön 4.3 11643 Twilight Erin Hunter 4.3 11643 First They Killed My Father: A Daughter of Cambodia Remembers Loung Ung 4.29 11643 First They Killed My Father: A Daughter of Cambodia Remembers Loung Ung 4.29 11644 Fool's Errand Robin Hobb 4.29 11643 Storia di chi fugge e di chi resta Elena Ferrante, Ann Goldstein 4.29 11643 Storia di chi fugge e di chi resta Elena Ferrante, Ann Goldstein 4

### **Record your thoughts about why you wrote the query the way you did.**

I started out by selecting all of the columns I would need to display the requested data. I decided on an inner join because it would get me the data I needed without any extraneous data (though all the join methods returned the same thing anyway in this case). Because there were null book titles, I set up a where statement to exclude them. I had to group by everything I selected since none of it was used in an aggregate function. For the having statement, I included only the tags I was looking for. I then ordered everything by average rating in descending order to show the highest rated first.

# Part 2: Choose Another Month

Choose another month and plan at least 2 events / promotions and answer the following questions.

Which month did you choose?  

What 2 events / promotions are you highlighting?

> 1. January

> 2. Houseplant Appreciation Day and International Holocaust Rememberance Day

# Part 3: Summarize your Work

For each event write at least one query that joins any two tables in the BooksDB to support your choice and record you thoughts as to why you used the paticlular query. At least one of your queries needs to include a HAVING clause.

In [42]:
-- Event 1 Query
select * from BooksDB.dbo.tags where tag_name like '%chlorophyll%' or tag_name like '%soil%' or tag_name like '%succulent%' or tag_name like '%plant%'

(16 rows affected)

Total execution time: 00:00:00.314

tag_id tag_name 4903 bone-marrow-transplant 7294 church-planting 8118 cook-books-plant-based 8341 cotton-plantations 14184 heart-transplant 23714 plant 23715 plant-based 23716 plantagenet 23717 plantagenets 23718 plantation 23719 plantation-life 23720 plantations 23721 planting 23722 plants 23723 plants-fruits-vegetables 31144 transplantation

In [40]:
select b.original_title as "Title", b.authors as "Authors", b.average_rating as "Average Rating", bt.tag_id as "Tag ID"
from BooksDB.dbo.book_tags as bt
inner join BooksDB.dbo.books as b
on b.best_book_id = bt.goodreads_book_id
group by b.original_title,b.authors,b.average_rating,bt.tag_id
having bt.tag_id like '23714' or bt.tag_id like '23721' or bt.tag_id like '23722' or bt.tag_id like '23723'
order by b.original_title asc;

(6 rows affected)

Total execution time: 00:00:01.163

Title Authors Average Rating Tag ID Lab Girl Hope Jahren 4.04 23722 The Botany of Desire: A Plant's-Eye View of the World Michael Pollan 4.05 23722 The Carrot Seed Ruth Krauss, Crockett Johnson 4.07 23721 The Carrot Seed Ruth Krauss, Crockett Johnson 4.07 23722 The Carrot Seed Ruth Krauss, Crockett Johnson 4.07 23723 The Day of the Triffids John Wyndham 4 23714

### **Summary of event 1:**

I first started by doing a search of the tags database for plant-related keywords. I selected the tag IDs I thought might apply to houseplants and used a query similar to the one used for women's history month for the same reasons as it was used for women's history month. All-in-all, this would be a very disappointing appreciation of houseplents. The only likely candidates are _The Botany of Desire: A Plant's-Eye View of the World_ and _The Carrot Seed_. Maybe we could sell some houseplants instead or cute houseplant themed bookmarks or knick knacks.

In [57]:
-- Event 2 Query
select * from BooksDB.dbo.tags where tag_name like '%holocaust%' or tag_name like '%wwii%' or tag_name like '%ww2%' or tag_name like '%nazi%';

(56 rows affected)

Total execution time: 00:00:00.363

tag_id tag_name 934 20th-century-post-ww2 11843 fiction-wwii 14352 hiding-jews-from-nazis 14512 historical-fiction-ww2 14513 historical-fiction-wwii 14627 history-us-wwii 14637 history-ww2 14639 history-wwii 14730 holocaust 14731 holocaust-and-wwii 14732 holocaust-books 14733 holocaust-fiction 14734 holocaust-genocide 14735 holocaust-lit 14736 holocaust-literature 14737 holocaust-memoir 14738 holocaust-memoirs 14739 holocaust-related 14740 holocaust-studies 14741 holocaust-survivor 14742 holocaust-survivors 14743 holocaust-ww2 14744 holocaust-wwii 14745 holocausto 16388 jewish-holocaust 21223 nazi 21224 nazi-french-occupation 21225 nazi-germany 21226 nazis 21227 nazism 25779 resistance-to-nazis 29850 the-holocaust 32157 war-ww2 32817 world-war-ii-holocaust 32906 ww2 32907 ww2-books 32908 ww2-fiction 32909 ww2-history 32910 ww2-holocaust 32911 ww2-pacific 32916 wwi-wwii 32917 wwii 32918 wwii-books 32919 wwii-collection 32920 wwii-era 32921 wwii-europe 32922 wwii-fiction 32923 wwii-historical-fiction 32924 wwii-history 32925 wwii-holocaust 32926 wwii-memoirs 32927 wwii-non-fiction 32928 wwii-nonfiction 32929 wwii-normandy 32930 wwii-pacific 32931 wwii-related

In [58]:
select b.original_title as "Title", b.authors as "Authors", b.average_rating as "Average Rating", bt.tag_id as "Tag ID"
from BooksDB.dbo.book_tags as bt
inner join BooksDB.dbo.books as b
on b.best_book_id = bt.goodreads_book_id
group by b.original_title,b.authors,b.average_rating,bt.tag_id
having bt.tag_id like '14352' or bt.tag_id like '14737' or bt.tag_id like '14742'
order by b.original_title asc;

(7 rows affected)

Total execution time: 00:00:01.346

Title Authors Average Rating Tag ID All But My Life Gerda Weissmann Klein 4.28 14737 All But My Life Gerda Weissmann Klein 4.28 14738 All But My Life Gerda Weissmann Klein 4.28 14741 Day After Night Anita Diamant 3.65 14742 I Have Lived A Thousand Years: Growing Up In The Holocaust Livia E. Bitton-Jackson 4.17 14737 The Paris Architect Charles Belfoure 3.98 14352 The Storyteller Jodi Picoult 4.26 14742

### **Summary of event 2:**

I started out by searching the tags database for holocaust related keywords. Since this is a day of rememberance, I selected the tag IDs that I felt focused more on the victims of the holocaust and those that aided them. I then queried for those IDs and found that 14738 and 14741 applied only to _All But My Life_, which was also covered by 14737, and so decided to rerun the query without those IDs to tighten up the code a bit. There aren't many books to choose from with my criteria, but they do seem to be rather highly rated, so this rememberance will be quality over quantity, I suppose.